tolstoy/src/syncer.rs

```diff
@@ -0,0 +1,283 @@
+
+pub struct Syncer {}
+
+struct UploadingTxReceiver<'c> {
+    pub tx_temp_uuids: HashMap<Entid, Uuid>,
+    pub is_done: bool,
+    remote_client: &'c RemoteClient,
+    remote_head: &'c Uuid,
+    rolling_temp_head: Option<Uuid>,
+}
+
+impl<'c> UploadingTxReceiver<'c> {
+    fn new(client: &'c RemoteClient, remote_head: &'c Uuid) -> UploadingTxReceiver<'c> {
+        UploadingTxReceiver {
+            tx_temp_uuids: HashMap::new(),
+            remote_client: client,
+            remote_head: remote_head,
+            rolling_temp_head: None,
+            is_done: false
+        }
+    }
+}
+
+impl<'c> TxReceiver for UploadingTxReceiver<'c> {
+    fn tx<T>(&mut self, tx_id: Entid, d: &mut T) -> Result<()>
+    where T: Iterator<Item=TxPart> {
+        // Yes, we generate a new UUID for a given Tx, even if we might
+        // already have one mapped locally. Pre-existing local mapping will
+        // be replaced if this sync succeeds entirely.
+        // If we're seeing this tx again, it implies that previous attempt
+        // to sync didn't update our local head. Something went wrong last time,
+        // and it's unwise to try to re-use these remote tx mappings.
+        // We just leave garbage txs to be GC'd on the server.
+        let tx_uuid = Uuid::new_v4();
+        self.tx_temp_uuids.insert(tx_id, tx_uuid);
+        let mut tx_chunks = vec![];
+
+        // TODO separate bits of network work should be combined into single 'future'
+
+        // Upload all chunks.
+        for datom in d {
+            let datom_uuid = Uuid::new_v4();
+            tx_chunks.push(datom_uuid);
+            self.remote_client.put_chunk(&datom_uuid, serde_json::to_string(&datom)?)?
+        }
+
+        // Upload tx.
+        // NB: At this point, we may choose to update remote & local heads.
+        // Depending on how much we're uploading, and how unreliable our connection
+        // is, this might be a good thing to do to ensure we make at least some progress.
+        // Comes at a cost of possibly increasing racing against other clients.
+        match self.rolling_temp_head {
+            Some(parent) => {
+                self.remote_client.put_transaction(&tx_uuid, &parent, &tx_chunks)?;
+                self.rolling_temp_head = Some(tx_uuid.clone());
+            },
+            None => self.remote_client.put_transaction(&tx_uuid, self.remote_head, &tx_chunks)?
+        }
+
+        Ok(())
+    }
+
+    fn done(&mut self) -> Result<()> {
+        self.is_done = true;
+        Ok(())
+    }
+}
+
+impl Syncer {
+    pub fn flow(sqlite: &mut rusqlite::Connection) -> Result<()> {
+        // Sketch of an upload flow:
+        // get remote head
+        // compare with local head
+        // if the same:
+        // - upload any local chunks, transactions
+        // - move server remote head
+        // - move local remote head
+        
+        // TODO configure this sync with some auth data
+        let username = "grisha-1";
```

Comment:
Is it possible to sync to multiple destinations? This is an interesting thing to write up.